### PR TITLE
Sign in issues

### PR DESCRIPTION
**Pull request by JPonte**
Sign in by phone should fill the field with email;
Tablet should start on register by email.
#### APK
[Download build #9449](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9449/artifact/build/artifact/wire-dev-PR1115-9449.apk)
[Download build #9453](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9453/artifact/build/artifact/wire-dev-PR1115-9453.apk)
[Download build #9459](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9459/artifact/build/artifact/wire-dev-PR1115-9459.apk)
[Download build #9461](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9461/artifact/build/artifact/wire-dev-PR1115-9461.apk)
[Download build #9468](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9468/artifact/build/artifact/wire-dev-PR1115-9468.apk)
[Download build #9470](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9470/artifact/build/artifact/wire-dev-PR1115-9470.apk)
[Download build #9472](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9472/artifact/build/artifact/wire-dev-PR1115-9472.apk)
[Download build #9477](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9477/artifact/build/artifact/wire-dev-PR1115-9477.apk)

### CODE REVIEW

**Comment by someoneEsle** [comment]
Bug 1
- user has email/password and phone number
- user taps sign in with phone number
- types phone number
- receives SMS, types it, taps next
- user is taken back to the login with email, email is prefilled correctly
- user taps sign in with phone again
- phone number is prefilled, user taps next
- endless spinning

**Comment by someoneEsle** [comment]
Bug 2
- user has email/password and phone number
- user taps sign in with phone number
- types phone number
- receives SMS, types it, taps next
- user is taken back to the login with email, email is prefilled correctly
- user signs in with a different email and password
- user can now see two users in the settings

**Comment by someoneEsle** [comment]
Bug 3
Forgot link password doesn't work

**Comment by JPonte** [comment]
build again

**Comment by JPonte** [comment]
build again

**Comment by JPonte** [comment]
build again

**Comment by someoneEsle** [comment]
Bug 4
user has email/password and phone number
user taps sign in with phone number
types phone number
receives SMS, types it, taps next
user sees screen with email and password
user taps "edit phone number"
conversations list opens

Bug 5
user taps "terms and conditions"
user taps back
user does not go back to registration screen

**Comment by someoneEsle** [comment]
Bug 6
- app has 3 accounts logged in
- log out of one
- try to add an account
- get stuck in an infinite loop

File: app/src/main/scala/com/waz/zclient/appentry/InsertPasswordFragment.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    signInController.isValid.onUi { setConfirmationButtonActive }
+    cancelButton.foreach(_.setOnClickListener(this))
+    forgotPassword.foreach(_.setOnClickListener(this))
+  }
+
+  private def setConfirmationButtonActive(active: Boolean): Unit = {
+    if(active)
+      confirmationButton.foreach(_.setState(PhoneConfirmationButton.State.CONFIRM))
+    else
+      confirmationButton.foreach(_.setState(PhoneConfirmationButton.State.NONE))
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/appentry/InsertPasswordFragment.scala]: `..._.setState(if (active) CONFIRM else NONE)`

File: app/src/main/scala/com/waz/zclient/appentry/InsertPasswordFragment.scala [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import com.waz.zclient.utils.ViewUtils
+import com.waz.zclient.{EntryError, FragmentHelper, OnBackPressedListener, R}
+
+class InsertPasswordFragment extends BaseFragment[Container] with FragmentHelper with View.OnClickListener with OnBackPressedListener {
+
+  lazy val signInController = inject[SignInController]
+
+  lazy val passwordTextWatch = TextListener(signInController.password ! _)
+
+  lazy val emailField = Option(findById[GuidedEditText](getView, R.id.email_field))
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/appentry/InsertPasswordFragment.scala]: seems as though these are only used within `onViewCreated` to set listeners and validators - you can probably get rid of the constructor variables, and they probably don't need to be options (but you can use `returning` if you want to keep them grouped nicely)

File: app/src/main/scala/com/waz/zclient/appentry/InsertPasswordFragment.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      false)
+
+  override def onBackPressed() = {
+    ZMessaging.currentAccounts.logout(false)
+    true
+  }
+}
+
+object InsertPasswordFragment {
+  val Tag = logTagFor[InsertPasswordFragment]
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/appentry/InsertPasswordFragment.scala]: `import com.waz.ZLog.ImplicitLogTag._`

File: app/src/main/scala/com/waz/zclient/utils/TextViewUtils.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.waz.zclient.utils
+
+import android.text.{Editable, TextWatcher}
+
+object TextViewUtils {
+  case class TextListener(callback: String => Unit) extends TextWatcher {
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/utils/TextViewUtils.scala]: Maybe you can add this class to the `com.waz.zclient.utils` package object?

Comment (deanrobertcook) [reply app/src/main/scala/com/waz/zclient/utils/TextViewUtils.scala]: Actually, you can add a `RichTextView` implicit class in the `utils` package, and add a method `addTextWatcher` or something like that.

**Comment by someoneEsle** [comment]
Bug 7
when user has email/password and phone number
user taps sign in with phone number
types phone number
receives SMS, types it, taps next
user sees screen with email and password
user taps "edit phone number"
*now I want to see the phone number I typed*

Bug 8
when user has email/password and phone number
user taps sign in with phone number
types phone number
receives SMS, types it, taps next
user sees screen with email and password
user taps "edit phone number"
user leaves the phone number unchanged, taps next
conversation list appears

Bug 9
after bug 8, the user tries to add the same account one more time (maybe this time they'll be successful). They type their phone number, get the sms, get the email screen, type the password, and successfully log in. However, if they go into the settings they see one too many account (the account they just logged in with is displayed twice, probably because the one from bug 8 is counted)

**Comment by someoneEsle** [comment]
Bug 11
asking to resend the SMS code shows the old warning that says "please log in with email and password"

Bug 12
user has email and password
user wants to sign in with phone, types phone, gets the sms, types the code
user goes back to edit phone number, then closes the login screen
user now wants to sign in again with phone number, types phone, gets the sms, types the code, taps next
At this point, either the app closes or you get endless spinning
Sometimes, if the user goes back to the settings page, they can see the profile picture of the new account but they cannot open it (nothing happens when tapping that icon)

**Comment by someoneEsle** [comment]
Bug 7 is not fixed

Bug 12 is not fixed

**Comment by someoneEsle** [comment]
Bug 13
log in with email.
add account with phone number.
now you have two accounts in the app.
Try to log in with phone number again (same phone number from the second account): after you type the phone number, tapping next will take you back to the conversation list and log out the second account you added.

**Comment by someoneEsle** [comment]
Bug 14
user has email and phone number
user types phone number, next, receives SMS, types the code, next
Now the user types the wrong password, gets a warning saying "wrong credentials" and is taken back to the initial login screen instead of being able to type their password again.

**Comment by someoneEsle** [comment]
We agreed with Meng that the last two bugs should not block this PR. I filed them, but this PR can be merged as it is since those bugs would require quite a bit of work on the sync engine.

**Pull request merged**